pkg/proxy/iptables/proxier.go

```diff
@@ -428,29 +440,87 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
 			// Deep-copy in case the service instance changes
 			info.loadBalancerStatus = *api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
 			info.sessionAffinityType = service.Spec.SessionAffinity
+			if l, ok := service.Annotations[apiservice.AnnotationExternalTraffic]; ok {
+				if l == apiservice.AnnotationValueExternalTrafficLocal {
+					info.onlyNodeLocalEndpoints = true
+				} else if l == apiservice.AnnotationValueExternalTrafficGlobal {
+					info.onlyNodeLocalEndpoints = false
+				} else {
+					glog.Errorf("Invalid value for annotation %v", apiservice.AnnotationExternalTraffic)
+					info.onlyNodeLocalEndpoints = false
+				}
+				if l, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; ok {
+					p, err := strconv.Atoi(l)
+					if err != nil {
+						glog.Errorf("Failed to parse annotation %v: %v",
+							apiservice.AnnotationHealthCheckNodePort, err)
+					}
+					info.healthCheckNodePort = p
+					healthcheck.AddServiceListener(serviceName.NamespacedName, info.healthCheckNodePort)
+				} else {
+					glog.Errorf("Service does not contain necessary annotation %v",
+						apiservice.AnnotationHealthCheckNodePort)
+				}
+			}
+
 			proxier.serviceMap[serviceName] = info
 
 			glog.V(4).Infof("added serviceInfo(%s): %s", serviceName, spew.Sdump(info))
 		}
 	}
 
 	staleUDPServices := sets.NewString()
-	// Remove services missing from the update.
-	for name := range proxier.serviceMap {
+	activeServiceListeners := sets.NewString()
```

Comment:
Discussed IRL, I think you can just use the original map and lookup 'name' below.